library/std/arithmetic_internal.qs

```diff
@@ -0,0 +1,163 @@
+            let (x0, xrest) = HeadAndRest(xs);
+            let (y0, yrest) = HeadAndRest(ys);
+
+            use carryOut = Qubit();
+            Carry(carryIn, x0, y0, carryOut);
+            AddWithCarryIn(carryOut, xrest, yrest);
+            Uncarry(carryIn, x0, y0, carryOut);
+        }
+    }
+
+    /// # Summary
+    /// Implements Half-adder. Adds qubit x to qubit y and sets carryOut appropriately
+    /// assuming carryOut is in |0> state.
+    @Config(Full)
+    internal operation HalfAdderInc(x : Qubit, y : Qubit, carryOut : Qubit) : Unit is Adj + Ctl {
+        body (...) {
+            ApplyAndWith0Target(x, y, carryOut);
+            CNOT(x, y);
+        }
+        adjoint auto;
+
+        controlled (ctls, ...) {
+            Fact(Length(ctls) == 1, "HalfAdder should be controlled by exactly one control qubit.");
+
+            let ctl = ctls[0];
+            use helper = Qubit();
+
+            within {
+                ApplyAndWith0Target(x, y, helper);
+            } apply {
+                ApplyAndWith0Target(ctl, helper, carryOut);
+            }
+            CCNOT(ctl, x, y);
+        }
+        controlled adjoint auto;
+    }
+
+    /// # Summary
+    /// Implements Full-adder. Adds qubit carryIn and x to qubit y and sets carryOut appropriately.
+    @Config(Full)
+    internal operation FullAdderInc(carryIn : Qubit, x : Qubit, y : Qubit, carryOut : Qubit) : Unit is Adj + Ctl {
+        body (...) {
+            // TODO: cannot use `Carry` operation here
+            CNOT(carryIn, x);
+            CNOT(carryIn, y);
+            CCNOT(x, y, carryOut);
+            CNOT(carryIn, carryOut);
+            CNOT(carryIn, x);
+            CNOT(x, y);
+        }
+        adjoint auto;
+
+        controlled (ctls, ...) {
+            Fact(Length(ctls) == 1, "FullAdder should be controlled by exactly one control qubit.");
+
+            let ctl = ctls[0];
+            use helper = Qubit();
+
+            Carry(carryIn, x, y, helper);
+            CCNOT(ctl, helper, carryOut);
+            Controlled Uncarry(ctls, (carryIn, x, y, helper));
+        }
+        controlled adjoint auto;
+    }
+
+    // Computes carryOut := carryIn + x + y
+    @Config(Full)
+    internal operation FullAdder(carryIn : Qubit, x : Qubit, y : Qubit, carryOut : Qubit) : Unit is Adj {
+        CNOT(x, y);
+        CNOT(x, carryIn);
+        ApplyAndWith0Target(y, carryIn, carryOut);
+        CNOT(x, y);
+        CNOT(x, carryOut);
+        CNOT(y, carryIn);
+    }
+
+    /// # Summary
+    /// Compute carry bit for a full adder.
+    @Config(Full)
+    internal operation Carry(carryIn : Qubit, x : Qubit, y : Qubit, carryOut : Qubit) : Unit is Adj + Ctl {
+        body (...) {
+            CNOT(carryIn, x);
+            CNOT(carryIn, y);
+            ApplyAndWith0Target(x, y, carryOut);
+            CNOT(carryIn, carryOut);
+        }
+        adjoint auto;
+        controlled (ctls, ...) {
+            // TODO: Is it not controlled actually?
```

Comment:
It's worth to point this out in fact. The `Carry` and `Uncarry` operations are meant to be used only in an in-place ripple-carry adder implementation. In the controlled ripple-carry adder it is sufficient to only control the `Uncarry` operation. By defining the `Carry` operation like this (i.e., having the controlled version being the same as the body), we can simplify the implementation of the ripple-carry adder, by automatically generating the controlled variant.

Reply:
I added the comment.